odl/tomo/backends/astra_setup.py

```diff
@@ -574,88 +588,90 @@ def astra_data(astra_geom, datatype, data=None, ndim=2, allow_copy=False):
-    # releases making the interface more coherent regarding CPU and GPU
-    type_map_cuda = {'parallel': 'cuda',  # I
-                     'parallel3d': 'cuda3d'}  # I
-    type_map_cuda['fanflat'] = type_map_cuda['parallel']
-    type_map_cuda['fanflat_vec'] = type_map_cuda['fanflat']
-    type_map_cuda['cone'] = type_map_cuda['parallel3d']
-    type_map_cuda['parallel3d_vec'] = type_map_cuda['parallel3d']
-    type_map_cuda['cone_vec'] = type_map_cuda['cone']
-
-    # create config dict
+    astra_geom = astra_proj_geom['type']
+    if (
+        astra_geom == 'parallel_vec'
+        and not astra_supports('par2d_vec_geometry')
+    ):
+        raise ValueError(
+            "'parallel_vec' geometry not supported in the current version "
+            "of ASTRA"
+        )
+
+    # Check if projector types are valid. We should not have to do this,
+    # but the errors from ASTRA are rather unspecific, so we check ourselves
+    # to know what's wrong.
+    astra_proj_type = str(astra_proj_type).lower()
+
+    if (
+        astra_proj_type == 'distance_driven'
+        and not astra_supports('par2d_distance_driven_proj')
+    ):
+        raise ValueError(
+            "'distance_driven' projector not supported in the current "
+            "version of ASTRA"
+        )
+
+    if astra_geom in {'parallel', 'parallel_vec'}:
+        valid_proj_types = ['line', 'linear', 'strip', 'cuda']
+        if astra_supports('par2d_distance_driven_proj'):
+            valid_proj_types.append('distance_driven')
+    elif astra_geom in {'fanflat', 'fanflat_vec'}:
+        valid_proj_types = ['line_fanflat', 'strip_fanflat', 'cuda']
+    elif astra_geom in {'parallel3d', 'parallel3d_vec'}:
+        valid_proj_types = ['linear3d', 'cuda3d']
+    elif astra_geom in {'cone', 'cone_vec'}:
+        valid_proj_types = ['linearcone', 'cuda3d']
+    else:
+        raise ValueError('invalid geometry type {!r}'.format(astra_geom))
+
+    if astra_proj_type not in valid_proj_types:
+        raise ValueError(
+            'projector type {!r} not in the set {} of valid types for '
+            'geometry type {!r}'
+            ''.format(astra_proj_type, valid_proj_types, astra_geom)
+        )
+
+    # Create config dict
     proj_cfg = {}
-    if impl == 'cpu':
-        proj_cfg['type'] = type_map_cpu[proj_type][vol_interp]
-    else:  # impl == 'cuda'
-        proj_cfg['type'] = type_map_cuda[proj_type]
+    proj_cfg['type'] = astra_proj_type
     proj_cfg['VolumeGeometry'] = astra_vol_geom
     proj_cfg['ProjectionGeometry'] = astra_proj_geom
     proj_cfg['options'] = {}
 
-    # Add the hacky 1/r^2 weighting exposed in intermediate versions of
-    # ASTRA
-    if (proj_type in ('cone', 'cone_vec') and
-            astra_supports('cone3d_hacky_density_weighting')):
+    # Add the hacky 1/r^2 weighting exposed in intermediate versions of ASTRA
+    # TODO(kohr-h): add non-hacky version when available
```

Comment:
Isn't this reputably less hacky now as per https://github.com/odlgroup/odl/issues/1415?

Reply:
No, I think the hackiness lies in the fact that it's not a proper 1/r^2 weighting. But I've forgotten the details.

Reply:
I guess the comment was:

> Furthermore, I've made the approximation of the GPU cone beam BP density weighting more accurate (at hopefully limited computational cost since the extra factor could actually be folded into existing constants, it turns out), and also adopted the 2D fan beam GPU code to match. This density weighting is now enabled by default.

Reply:
I see. But I guess a true distance weighting would differ per location in the volume, which I don't think ASTRA implements. Likely it's only a function of detector position.

Reply:
I checked the quote, it refers to a work-in-progress branch, so we can be fairly sure but not 100 % that the next release will contain the improved weighting.